Lobby Player Limit

SaddestUnderscore opened the pull request:
This adds a lobby limit as an option that is configurable in the Lobby select screen
- Within a minimum of 2 and a maximum of 32 players
- Adds a little text where it also tells the max of the lobby in the bottom right of lobby cards

6fears7 commented:
Once you believe you're good to go to have it tested, please set the PR to ready for review.

SaddestUnderscore commented:
Hopefully this should be good to go, slight disclaimers though:
- testing in local won't do anything since this is more meant for online, plus the fact that i can't get it to work in local anyway
- i made the check for playercount to accept 1 player and no more, if the test is successful i will put the limit to 4 which is the normal

Chrometurtle commented on Online/Matchmaking/SteamMatchmakingManager.cs:
```diff
@@ -120,7 +121,24 @@ public override void CreateLobby(LobbyVisibility visibility, string gameMode)
-            m_JoinLobbyCall.Set(SteamMatchmaking.JoinLobby(lobby.id));
+            if (lobby.mode == "Story")
+            {
+                if (lobby.playerCount > 1)
```
The limit should be set from a UI element, not hardcoded

SaddestUnderscore replied:
yeah you are right, i do want to attempt making the limit configurable, but for now it's hardcoded for the meantime (until i figure out making it configurable)

SaddestUnderscore commented on Online/Matchmaking/SteamMatchmakingManager.cs:
```diff
@@ -120,7 +121,24 @@ public override void CreateLobby(LobbyVisibility visibility, string gameMode)
         public override void JoinLobby(LobbyInfo lobby)
         {
-            m_JoinLobbyCall.Set(SteamMatchmaking.JoinLobby(lobby.id));
+            if (lobby.mode == "Story")
```
like arena and freeroam right

6fears7 commented:
We can using a Debug build whenever you're free next.

SaddestUnderscore commented:
i'll list whats changed in this commit: added an option for the player to set the player max in the form of textbox, currently doesn't do anything yet, moved meadow display to the bottom so it doesn't interfere with the coords once everything is done and merged (including passwords where coords), scratched lobby mode detection

Chrometurtle commented on Menu/LobbySelectMenu.cs:
```diff
@@ -89,6 +87,20 @@ public LobbySelectMenu(ProcessManager manager) : base(manager, RainMeadow.Ext_Pr
+            where.y += 5; 
+
+            // display version
+            where = new Vector2(0f, 0f);
```
Bad, don't do this

SaddestUnderscore commented:
Okay, after lots of pain and debugging, this _Should_ work, I think I'm actually ready now for making this ready for review

6fears7 commented:
@Chrometurtle : 

Functionally, this is ready to go. When a lobby is full, the user is prompted, denied entry, and will safely return back to the main menu and can make their own game. 

When lobby has room, will join.

Picture of popup prompt:
![image](https://github.com/henpemaz/Rain-Meadow/assets/57415489/24abcc50-395c-47c8-a203-44cdcd420fd1)


Please check the code logic at your earliest convenience. In terms of stability, it seems warranted to merge to main.

Chrometurtle commented on Online/Resource/Lobby.cs:
```diff
@@ -15,10 +15,10 @@ public class Lobby : OnlineResource
         public string[] mods = RainMeadowModManager.GetActiveMods();
         public static bool modsChecked;
-
+        public int maxPlayerCount;
```
Looks like everything is already self contained in LobbyInfo. So I don't think this isn't needed anymore

Chrometurtle commented on Online/Matchmaking/SteamMatchmakingManager.cs:
```diff
@@ -8,6 +8,7 @@ namespace RainMeadow
 {
     public class SteamMatchmakingManager : MatchmakingManager
     {
+        public static int maxLobbyCount;
```
Same with this guy

SaddestUnderscore commented on Online/Matchmaking/MatchmakingManager.cs:
```diff
@@ -13,6 +13,7 @@ public abstract class MatchmakingManager
         public static string NAME_KEY = "name";
         public static string MODE_KEY = "mode";
         public static string PASSWORD_KEY = "password";
+        public static int MAX_LOBBY = 4;
```
MAX_LOBBY is needed, maxLobbyCount no

Chrometurtle commented:
lgtm, there is a way to simplify MAX_LOBBY ;3 but otherwise looks good. Merging
